File: boj/MinimumSpanningTree/BOJ_2887.py
# 백준 2887 행성 터널
# 21.03.15
import sys
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
rd = sys.stdin.readline
INF = int(1e9)

def find(a):
    if parent[a] == a: return a
    else:
        parent[a] = find(parent[a])
        return parent[a]


def union(a, b):
    x = find(a)
    y = find(b)
    if x < y: parent[y] = x
    else: parent[x] = y

n = int(input())
parent = [0] * n
for i in range(n): parent[i] = i
edges = []
x = []
y = []
z = []
for i in range(n):
    a, b, c = map(int, rd().split())
    x.append((a, i))
    y.append((b, i))
    z.append((c, i))

x.sort(reverse=True)
y.sort(reverse=True)
z.sort(reverse=True)

for i in range(1, n):
    edges.append((abs(x[i][0] - x[i-1][0]), x[i-1][1], x[i][1]))
    edges.append((abs(y[i][0] - y[i-1][0]), y[i-1][1], y[i][1]))
    edges.append((abs(z[i][0] - z[i-1][0]), z[i-1][1], z[i][1]))

edges.sort(reverse=True)
cnt = 0
ans = 0
while cnt < n - 1:
    w, u, v = edges.pop()

    if find(u) != find(v):
        union(u, v)
        cnt += 1
        ans += w

print(ans)

File: boj/MinimumSpanningTree/BOJ_2887.swift
// 백준 2887 행성 터널
// 21.03.17

import Foundation

func find (a: Int) -> Int {
  if parent[a] != a {
    parent[a] = find(a: parent[a])
    return parent[a]
  }
  else { return a }
}

func union (a: Int, b: Int) {
  let x = find(a: a)
  let y = find(a: b)
  
  if x < y { parent[y] = x }
  else { parent[x] = y }
}

let n = Int(readLine()!)!
var parent = Array(repeating: 0, count: n + 1)
var x: [(Int, Int)] = []
var y: [(Int, Int)] = []
var z: [(Int, Int)] = []

for i in stride(from: 1, through: n, by: 1) {
  let temp = readLine()!.split(separator: " ").map { Int(String($0))! }
  x.append((temp[0], i))
  y.append((temp[1], i))
  z.append((temp[2], i))
  parent[i] = i
}

x.sort(by: { $0.0 < $1.0 })
y.sort(by: { $0.0 < $1.0 })
z.sort(by: { $0.0 < $1.0 })

var edges: [(Int, Int, Int)] = []
for i in stride(from: 1, to: n, by: 1) {
  edges.append((abs(x[i].0 - x[i-1].0), x[i].1, x[i-1].1))
  edges.append((abs(y[i].0 - y[i-1].0), y[i].1, y[i-1].1))
  edges.append((abs(z[i].0 - z[i-1].0), z[i].1, z[i-1].1))
}

edges.sort(by: { $0.0 > $1.0 })
var sum = 0

while !edges.isEmpty {
  let pop = edges.removeLast()
  let w = pop.0
  let u = pop.1
  let v = pop.2
  
  if find(a: u) != find(a: v) {
    union(a: u, b: v)
    sum += w
  }
}

print(sum)
